proposals/custom-page-sizes/Overview.md

```diff
@@ -1,3 +1,294 @@
+
+[Memory type matching]: https://webassembly.github.io/spec/core/valid/types.html#memories
+
+The `memory.grow` and `memory.size` instructions continue to give results in
+page counts, and can generally remain unmodified.
+
+Customizing a memory's page size does not affect its index type; it has the same
+`i32` or `i64`[^i64-index] index it would otherwise have.
+
+[^i64-index]: If the `memory64` proposal is enabled and this memory is a 64-bit
+memory.
+
+Finally, extending memory types with a page size has the following desirable
+properties:
+
+1. **It is minimally invasive.** This proposal doesn't require any new
+   instructions (e.g. a `memory.page_size <memidx> : [] -> [u32]` instruction,
+   which would be required if page sizes were not static knowledge, see point 3)
+   and only imposes small adjustments to existing instructions. For example, the
+   `memory.size` instruction, as mentioned previously, is almost unchanged: its
+   result is still defined in terms of numbers of pages, its just that engines
+   divide the memory's byte capacity by its defined page size rather than the
+   constant 64KiB.
+
+2. **Page size customization applies to a particular memory.** A module is free
+   to, for example, define multiple memories with multiple different page
+   sizes. Opting into small page sizes for one memory does not affect any other
+   memories in the store.
+
+3. **Page sizes are always known statically.** Every memory instruction has a
+   static `<memidx>` immediate, so we statically know the memory it is operating
+   upon as well as that memory's type, which means we additionally have static
+   knowledge of the memory's page size. This enables constant-propagation and
+   -folding optimizations in Wasm producers and engines with ahead-of-time
+   compilers.
+
+### Example
+
+Here is a short example using strawperson WAT syntax:
+
+```wat
+(module
+  ;; Import a memory with a page size of 512 bytes and a minimum size of
+  ;; 2 pages, aka 1024 bytes. No maximum is specified.
+  (import "env" "memory" (memory $imported (page_size 512) 2))
+
+  ;; Define a memory with a page size of 1; a minimum size of 13 pages, aka
+  ;; 13 bytes; and a maximum size of 42 pages, aka 42 bytes.
+  (memory $defined (page_size 1) 13 42)
+
+  ;; Export a function to get the imported memory's size, in bytes.
+  (func (export "get_imported_memory_size_in_bytes") (result i32)
+    ;; Get the current size of the memory in units of pages.
+    memory.size $imported
+    ;; And thenmultiply by the bytes-per-page to get the memory's byte size.
+    i32.const 512
+    i32.mul
+  )
+
+  ;; And export a similar function for the defined memory. In this case we can
+  ;; avoid the multiplication by page size, since we statically know the page
+  ;; size is 1.
+  (func (export "get_defined_memory_size_in_bytes") (result i32)
+    memory.size $defined
+  )
+)
+```
+
+### Binary Encoding
+
 TODO
```

Comment:
Not relevant to this PR, but for the future PR to fill this in I found [this table in the memory64 proposal](https://github.com/WebAssembly/memory64/blob/main/proposals/memory64/Overview.md#binary-format) perhaps a good starting point for determining a binary encoding. (aka pick the next bit in the prefix byte and have it represent a "trailing log2 page size i32" and that may be all that's necessary)

Reply:
Thanks for the pointer and the reminder to file an issue. Just did so: https://github.com/WebAssembly/custom-page-sizes/issues/5